File: ui-tests/PageObjects/Panes/Terminal/TerminalOutputPane.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects;

namespace UiTests.PageObjects.Panes.Terminal;

/// <summary>
/// Wraps the terminal output component within the layout.
/// </summary>
public class TerminalOutputPane : TabObject
{
    private List<TerminalLine> _lines = new();

    public TerminalOutputPane(IPage page, ILocator root, string tabButtonText)
        : base(page, root, tabButtonText)
    {
    }

    /// <summary>
    /// Locator for the &lt;pre&gt; block containing terminal lines.
    /// </summary>
    public ILocator LinesContainer()
        => Root.Locator("pre");

    /// <summary>
    /// Returns the currently rendered terminal lines.
    /// </summary>
    public async Task<IReadOnlyList<TerminalLine>> LinesAsync(bool forceReload = false)
    {
        if (forceReload || _lines.Count == 0)
        {
            var roots = await LinesContainer().Locator(".terminal-line").AllAsync();
            _lines = roots.Select(locator => new TerminalLine(locator)).ToList();
        }

        return _lines;
    }

    /// <summary>
    /// Provides direct access to a terminal line by index.
    /// </summary>
    public async Task<TerminalLine?> LineByIndexAsync(int index, bool forceReload = false)
    {
        var lines = await LinesAsync(forceReload);
        foreach (var line in lines)
        {
            if (await line.LineIndexAsync() == index)
            {
                return line;
            }
        }

        return null;
    }
}

File: tsc-ui-tests/page-objects/panes/terminal/terminal-output-pane.ts
import type { Locator, Page } from "@playwright/test";
import { TerminalLine } from "./terminal-line";

/**
 * Wraps the terminal output component within the layout.
 *
 * Port of ui-tests/PageObjects/Panes/Terminal/TerminalOutputPane.cs
 */
export class TerminalOutputPane {
  readonly page: Page;
  readonly root: Locator;
  readonly tabButtonText: string;
  private lines_: TerminalLine[] = [];

  constructor(page: Page, root: Locator, tabButtonText: string) {
    this.page = page;
    this.root = root;
    this.tabButtonText = tabButtonText;
  }

  tabButton(): Locator {
    return this.page.locator(".lm_title", { hasText: this.tabButtonText }).first();
  }

  linesContainer(): Locator {
    return this.root.locator("pre");
  }

  async lines(forceReload = false): Promise<TerminalLine[]> {
    if (forceReload || this.lines_.length === 0) {
      const roots = await this.linesContainer().locator(".terminal-line").all();
      this.lines_ = roots.map((l) => new TerminalLine(l));
    }
    return this.lines_;
  }

  async lineByIndex(
    index: number,
    forceReload = false,
  ): Promise<TerminalLine | null> {
    const lines = await this.lines(forceReload);
    for (const line of lines) {
      if ((await line.lineIndex()) === index) {
        return line;
      }
    }
    return null;
  }
}
